File: android/app/src/main/java/com/tanay/echo/transcription/VoiceCommands.kt
package com.tanay.echo.transcription

/**
 * Spoken formatting commands: paragraph/line breaks and named punctuation become their literal
 * characters. Deterministic and instant, so core English commands still work without AI cleanup.
 * Mirrors desktop src/shared/voice-commands.ts.
 */

private const val PARAGRAPH_PHRASES =
    """(?:start|begin|make)\s+(?:a\s+)?new\s+paragraph|(?:new|next)\s+paragraph|paragraph\s+break|blank\s+line|leave\s+(?:a\s+)?(?:space|gap)"""
private const val LINE_PHRASES = """(?:new|next)\s+line|line\s+break"""
private const val PUNCTUATION_PHRASES =
    """full\s+stop|period|comma|question\s+mark|exclamation\s+(?:mark|point)|colon|semi[ -]?colon|ellipsis|dot\s+dot\s+dot|hyphen|(?:open|left)\s+parenthes(?:is|es)|(?:close|right)\s+parenthes(?:is|es)|(?:open|begin|start)\s+quote|(?:close|end)\s+quote"""

// Optional leading article ("a new paragraph about X") marks the phrase as content, not a command.
private val BREAK_COMMAND = Regex(
    """(\b(?:a|the|one)\s+)?\b($PARAGRAPH_PHRASES|$LINE_PHRASES)\b[.,!?;:]*\s*""",
    RegexOption.IGNORE_CASE,
)
private val PUNCTUATION_COMMAND = Regex(
    """(\b(?:a|the|one)\s+)?\b($PUNCTUATION_PHRASES)\b[.,!?;:]*[ \t]*""",
    RegexOption.IGNORE_CASE,
)

private const val MARK_COMMA = "\uE000"
private const val MARK_PERIOD = "\uE001"
private const val MARK_QUESTION = "\uE002"
private const val MARK_EXCLAMATION = "\uE003"
private const val MARK_COLON = "\uE004"
private const val MARK_SEMICOLON = "\uE005"
private const val MARK_ELLIPSIS = "\uE006"
private const val MARK_HYPHEN = "\uE007"
private const val MARK_OPEN_PAREN = "\uE008"
private const val MARK_CLOSE_PAREN = "\uE009"
private const val MARK_OPEN_QUOTE = "\uE00A"
private const val MARK_CLOSE_QUOTE = "\uE00B"

private val SPACE_BEFORE_BREAK = Regex("""[ \t]+\n""")
private val EXTRA_BREAKS = Regex("""\n{3,}""")
private val LOWER_AFTER_BREAK = Regex("""\n([a-z])""")
private val LOWER_AFTER_TERMINAL = Regex("""([.!?…]) ([a-z])""")
private val LEADING_BREAKS = Regex("""^\n+""")
private val TRAILING_WHITESPACE = Regex("""\s+$""")
private val TRAILING_LINE_WHITESPACE = Regex("""[ \t]+$""", RegexOption.MULTILINE)

/**
 * Replace spoken formatting commands in a transcript with real breaks: paragraph commands become a
 * blank line, line commands a single newline. Whisper's own punctuation around the command is
 * absorbed, the first word after a break is capitalized, and phrases preceded by an article
 * ("add a new paragraph about…") are left literal. Total — never throws.
 */
fun applyVoiceCommands(text: String): String {
    if (text.isEmpty()) return text

    var replaced = BREAK_COMMAND.replace(text) { m ->
        val article = m.groupValues[1]
        val phrase = m.groupValues[2]
        when {
            article.isNotEmpty() -> m.value
            phrase.contains("line", ignoreCase = true) && !phrase.contains("blank", ignoreCase = true) -> "\n"
            else -> "\n\n"
        }
    }

    val source = replaced
    replaced = PUNCTUATION_COMMAND.replace(replaced) { m ->
        val article = m.groupValues[1]
        val phrase = m.groupValues[2]
        if (article.isNotEmpty() || isClearlyLiteralPunctuation(m.range.first, m.value, source)) {
            m.value
        } else {
            punctuationMark(phrase)
        }
    }

    return replaced
        .replace(Regex("""[ \t]*$MARK_COMMA[ \t]*"""), ", ")
        .replace(Regex("""[ \t]*$MARK_PERIOD[ \t]*"""), ". ")
        .replace(Regex("""[ \t]*$MARK_QUESTION[ \t]*"""), "? ")
        .replace(Regex("""[ \t]*$MARK_EXCLAMATION[ \t]*"""), "! ")
        .replace(Regex("""[ \t]*$MARK_COLON[ \t]*"""), ": ")
        .replace(Regex("""[ \t]*$MARK_SEMICOLON[ \t]*"""), "; ")
        .replace(Regex("""[ \t]*$MARK_ELLIPSIS[ \t]*"""), "… ")
        .replace(Regex("""[ \t]*$MARK_HYPHEN[ \t]*"""), " - ")
        .replace(Regex("""[ \t]*$MARK_OPEN_PAREN[ \t]*"""), " (")
        .replace(Regex("""[ \t]*$MARK_CLOSE_PAREN"""), ")")
        .replace(Regex("""[ \t]*$MARK_OPEN_QUOTE[ \t]*"""), " \"")
        .replace(Regex("""[ \t]*$MARK_CLOSE_QUOTE"""), "\"")
        .replace(Regex("""([,;:])\1+"""), "$1")
        .replace(Regex("""([.!?…])[.!?…]+"""), "$1")
        .replace(Regex("""[,;:]+([.!?…])"""), "$1")
        .replace(Regex("""[ \t]{2,}"""), " ")
        .replace(LOWER_AFTER_TERMINAL) { "${it.groupValues[1]} ${it.groupValues[2].uppercase()}" }
        .replace(SPACE_BEFORE_BREAK, "\n")
        .replace(EXTRA_BREAKS, "\n\n")
        .replace(LOWER_AFTER_BREAK) { "\n" + it.groupValues[1].uppercase() }
        .replace(TRAILING_LINE_WHITESPACE, "")
        .replace(Regex("""^[ \t]+"""), "")
        .replace(LEADING_BREAKS, "")
        .replace(TRAILING_WHITESPACE, "")
}

private fun punctuationMark(phrase: String): String {
    val normalized = phrase.lowercase().replace(Regex("[ -]+"), " ")
    return when {
        normalized == "comma" -> MARK_COMMA
        normalized == "full stop" || normalized == "period" -> MARK_PERIOD
        normalized == "question mark" -> MARK_QUESTION
        normalized.startsWith("exclamation ") -> MARK_EXCLAMATION
        normalized == "colon" -> MARK_COLON
        normalized == "semicolon" || normalized == "semi colon" -> MARK_SEMICOLON
        normalized == "ellipsis" || normalized == "dot dot dot" -> MARK_ELLIPSIS
        normalized == "hyphen" -> MARK_HYPHEN
        Regex("""^(?:open|left) parenthes""").containsMatchIn(normalized) -> MARK_OPEN_PAREN
        Regex("""^(?:close|right) parenthes""").containsMatchIn(normalized) -> MARK_CLOSE_PAREN
        Regex("""^(?:open|begin|start) quote$""").matches(normalized) -> MARK_OPEN_QUOTE
        else -> MARK_CLOSE_QUOTE
    }
}

private fun isClearlyLiteralPunctuation(offset: Int, match: String, source: String): Boolean {
    val before = source.take(offset)
    val remainder = source.drop(offset + match.length)
    if (Regex("""^-?(?:separated|delimited|operator|character|key|symbol|means|is|was|ended|began|ends|lasts|of|for|between|cancer)\b""", RegexOption.IGNORE_CASE)
            .containsMatchIn(remainder)
    ) return true
    if (Regex("""\b(?:word|term|character|symbol|key|literal|oxford|trial|billing|grace|menstrual|historical|time)\s+$""", RegexOption.IGNORE_CASE)
            .containsMatchIn(before)
    ) return true
    return offset == 0 && Regex("""^(?:means|is|was|ended|began)\b""", RegexOption.IGNORE_CASE)
        .containsMatchIn(remainder)
}

File: src/shared/voice-commands.ts
// Spoken formatting commands: paragraph/line breaks and named punctuation become their literal
// characters. Deterministic and instant: applied after dictionary correction and before AI cleanup,
// so core English dictation commands still work when cleanup is unavailable.

const PARAGRAPH_PHRASES =
  String.raw`(?:start|begin|make)\s+(?:a\s+)?new\s+paragraph|` +
  String.raw`(?:new|next)\s+paragraph|paragraph\s+break|blank\s+line|leave\s+(?:a\s+)?(?:space|gap)`
const LINE_PHRASES = String.raw`(?:new|next)\s+line|line\s+break`
const PUNCTUATION_PHRASES =
  String.raw`full\s+stop|period|comma|question\s+mark|` +
  String.raw`exclamation\s+(?:mark|point)|colon|semi[ -]?colon|ellipsis|dot\s+dot\s+dot|` +
  String.raw`hyphen|(?:open|left)\s+parenthes(?:is|es)|(?:close|right)\s+parenthes(?:is|es)|` +
  String.raw`(?:open|begin|start)\s+quote|(?:close|end)\s+quote`

// Optional leading article ("a new paragraph about X") marks the phrase as content, not a command.
const BREAK_COMMAND = new RegExp(
  String.raw`(\b(?:a|the|one)\s+)?\b(${PARAGRAPH_PHRASES}|${LINE_PHRASES})\b[.,!?;:]*\s*`,
  'gi'
)
const PUNCTUATION_COMMAND = new RegExp(
  String.raw`(\b(?:a|the|one)\s+)?\b(${PUNCTUATION_PHRASES})\b[.,!?;:]*[ \t]*`,
  'gi'
)

const MARK = {
  comma: '\uE000',
  period: '\uE001',
  question: '\uE002',
  exclamation: '\uE003',
  colon: '\uE004',
  semicolon: '\uE005',
  ellipsis: '\uE006',
  hyphen: '\uE007',
  openParen: '\uE008',
  closeParen: '\uE009',
  openQuote: '\uE00A',
  closeQuote: '\uE00B'
} as const

/**
 * Replace spoken formatting commands with real breaks and punctuation. Whisper's punctuation
 * around the command is absorbed, the next word is capitalized, and clearly literal uses remain
 * ordinary prose. Total: never throws, and no-command text passes through untouched.
 */
export function applyVoiceCommands(text: string): string {
  if (!text) return text

  let replaced = text.replace(
    BREAK_COMMAND,
    (match, article: string | undefined, phrase: string) => {
      if (article) return match
      return /line/i.test(phrase) && !/blank/i.test(phrase) ? '\n' : '\n\n'
    }
  )

  replaced = replaced.replace(
    PUNCTUATION_COMMAND,
    (match, article: string | undefined, phrase: string, offset: number, source: string) => {
      if (article || isClearlyLiteralPunctuation(offset, match, source)) return match
      return punctuationMark(phrase)
    }
  )

  return replaced
    .replace(new RegExp(`[ \\t]*${MARK.comma}[ \\t]*`, 'g'), ', ')
    .replace(new RegExp(`[ \\t]*${MARK.period}[ \\t]*`, 'g'), '. ')
    .replace(new RegExp(`[ \\t]*${MARK.question}[ \\t]*`, 'g'), '? ')
    .replace(new RegExp(`[ \\t]*${MARK.exclamation}[ \\t]*`, 'g'), '! ')
    .replace(new RegExp(`[ \\t]*${MARK.colon}[ \\t]*`, 'g'), ': ')
    .replace(new RegExp(`[ \\t]*${MARK.semicolon}[ \\t]*`, 'g'), '; ')
    .replace(new RegExp(`[ \\t]*${MARK.ellipsis}[ \\t]*`, 'g'), '… ')
    .replace(new RegExp(`[ \\t]*${MARK.hyphen}[ \\t]*`, 'g'), ' - ')
    .replace(new RegExp(`[ \\t]*${MARK.openParen}[ \\t]*`, 'g'), ' (')
    .replace(new RegExp(`[ \\t]*${MARK.closeParen}`, 'g'), ')')
    .replace(new RegExp(`[ \\t]*${MARK.openQuote}[ \\t]*`, 'g'), ' "')
    .replace(new RegExp(`[ \\t]*${MARK.closeQuote}`, 'g'), '"')
    .replace(/([,;:])\1+/g, '$1')
    .replace(/([.!?…])[.!?…]+/g, '$1')
    .replace(/[,;:]+([.!?…])/g, '$1')
    .replace(/[ \t]{2,}/g, ' ')
    .replace(/([.!?…]) ([a-z])/g, (_m, punctuation: string, ch: string) => punctuation + ' ' + ch.toUpperCase())
    .replace(/[ \t]+\n/g, '\n')
    .replace(/\n{3,}/g, '\n\n')
    .replace(/\n([a-z])/g, (_m, ch: string) => '\n' + ch.toUpperCase())
    .replace(/[ \t]+$/gm, '')
    .replace(/^[ \t]+/, '')
    .replace(/^\n+/, '')
    .replace(/\s+$/, '')
}

function punctuationMark(phrase: string): string {
  const normalized = phrase.toLowerCase().replace(/[ -]+/g, ' ')
  if (normalized === 'comma') return MARK.comma
  if (normalized === 'full stop' || normalized === 'period') return MARK.period
  if (normalized === 'question mark') return MARK.question
  if (normalized.startsWith('exclamation ')) return MARK.exclamation
  if (normalized === 'colon') return MARK.colon
  if (normalized === 'semicolon' || normalized === 'semi colon') return MARK.semicolon
  if (normalized === 'ellipsis' || normalized === 'dot dot dot') return MARK.ellipsis
  if (normalized === 'hyphen') return MARK.hyphen
  if (/^(?:open|left) parenthes/.test(normalized)) return MARK.openParen
  if (/^(?:close|right) parenthes/.test(normalized)) return MARK.closeParen
  if (/^(?:open|begin|start) quote$/.test(normalized)) return MARK.openQuote
  return MARK.closeQuote
}

function isClearlyLiteralPunctuation(offset: number, match: string, source: string): boolean {
  const before = source.slice(0, offset)
  const remainder = source.slice(offset + match.length)
  if (
    /^-?(?:separated|delimited|operator|character|key|symbol|means|is|was|ended|began|ends|lasts|of|for|between|cancer)\b/i.test(
      remainder
    )
  ) {
    return true
  }
  if (
    /\b(?:word|term|character|symbol|key|literal|oxford|trial|billing|grace|menstrual|historical|time)\s+$/i.test(
      before
    )
  ) {
    return true
  }
  return offset === 0 && /^(?:means|is|was|ended|began)\b/i.test(remainder)
}
